src/main/scala/io/iohk/ethereum/domain/BlockHeader.scala

```diff
@@ -52,71 +79,166 @@ case class BlockHeader(
+  )
+
+  def buildPostECIP1098Header(parentHash: ByteString,
+                              ommersHash: ByteString,
+                              beneficiary: ByteString,
+                              stateRoot: ByteString,
+                              transactionsRoot: ByteString,
+                              receiptsRoot: ByteString,
+                              logsBloom: ByteString,
+                              difficulty: BigInt,
+                              number: BigInt,
+                              gasLimit: BigInt,
+                              gasUsed: BigInt,
+                              unixTimestamp: Long,
+                              extraData: ByteString,
+                              mixHash: ByteString,
+                              nonce: ByteString,
+                              treasuryOptOut: Boolean): BlockHeader = BlockHeader(
+    parentHash = parentHash,
+    ommersHash = ommersHash,
+    beneficiary = beneficiary,
+    stateRoot = stateRoot,
+    transactionsRoot = transactionsRoot,
+    receiptsRoot = receiptsRoot,
+    logsBloom = logsBloom,
+    difficulty = difficulty,
+    number = number,
+    gasLimit = gasLimit,
+    gasUsed = gasUsed,
+    unixTimestamp = unixTimestamp,
+    extraData = extraData,
+    mixHash = mixHash,
+    nonce = nonce,
+    extraFields = HefPostEcip1098(treasuryOptOut)
+  )
+
+  def buildPostECIP1097Header(parentHash: ByteString,
+                              ommersHash: ByteString,
+                              beneficiary: ByteString,
+                              stateRoot: ByteString,
+                              transactionsRoot: ByteString,
+                              receiptsRoot: ByteString,
+                              logsBloom: ByteString,
+                              difficulty: BigInt,
+                              number: BigInt,
+                              gasLimit: BigInt,
+                              gasUsed: BigInt,
+                              unixTimestamp: Long,
+                              extraData: ByteString,
+                              mixHash: ByteString,
+                              nonce: ByteString,
+                              treasuryOptOut: Boolean,
+                              checkpoint: Option[Checkpoint]): BlockHeader = BlockHeader(
+    parentHash = parentHash,
+    ommersHash = ommersHash,
+    beneficiary = beneficiary,
+    stateRoot = stateRoot,
+    transactionsRoot = transactionsRoot,
+    receiptsRoot = receiptsRoot,
+    logsBloom = logsBloom,
+    difficulty = difficulty,
+    number = number,
+    gasLimit = gasLimit,
+    gasUsed = gasUsed,
+    unixTimestamp = unixTimestamp,
+    extraData = extraData,
+    mixHash = mixHash,
+    nonce = nonce,
+    extraFields = HefPostEcip1097(treasuryOptOut, checkpoint)
+  )
+  // scalastyle:on parameter.number
+
+  sealed trait HeaderExtraFields
+  object HeaderExtraFields {
+    case object HefPreEcip1098 extends HeaderExtraFields
```

Comment:
I think a better name would be `HefEmpty` - it would immediately stand out from the others and be associated with the default case, whereas the current name requires some attention to be distinguished from others.